ephyviewer/datasource/epochs.py

```diff
@@ -54,152 +55,219 @@ class WritableEpochSource(InMemoryEpochSource):
     """
-    def __init__(self, epoch, possible_labels, color_labels=None):
-        InMemoryEpochSource.__init__(self, all_epochs=[epoch])
+    def __init__(self, epoch=None, possible_labels=[], color_labels=None, channel_name=''):
         
-        #~ self._t_stop = max([ np.max(e['time']+e['duration']) for e in self.all if len(e['time'])>0])
+        self.possible_labels = possible_labels
+        self.channel_name = channel_name
         
-        self.times = self.all[0]['time']
-        self.durations = self.all[0]['duration']
+        if epoch is None:
+            epoch = self.load()
+        
+        InMemoryEpochSource.__init__(self, all_epochs=[epoch])
+        
+        # assign each epoch a fixed, unique integer id
+        self._next_id = 0
+        for chan in self.all:
+            chan['id'] = np.arange(self._next_id, self._next_id + len(chan['time']))
+            self._next_id += len(chan['time'])
         
         assert self.all[0]['time'].dtype.kind=='f'
         assert self.all[0]['duration'].dtype.kind=='f'
-        assert np.all((self.times[:-1]+self.durations[:-1])<=self.times[1:])
+        assert np.all(np.in1d(epoch['label'], self.possible_labels))
         
-        assert np.all(np.in1d(epoch['label'], possible_labels))
-        self.possible_labels = possible_labels
+        # put the epochs into a canonical order after loading
+        self._clean_and_set(self.all[0]['time'], self.all[0]['duration'], self.all[0]['label'], self.all[0]['id'])
         
         if color_labels is None:
-            n = len(possible_labels)
+            n = len(self.possible_labels)
             cmap = matplotlib.cm.get_cmap('Dark2' , n)
             color_labels = [ matplotlib.colors.ColorConverter().to_rgb(cmap(i)) for i in  range(n)]
             color_labels = (np.array(color_labels)*255).astype(int)
             color_labels = color_labels.tolist()
         self.color_labels = color_labels
         self.label_to_color = dict(zip(self.possible_labels, self.color_labels))
         
+    @property
+    def ep_times(self):
+        return self.all[0]['time']
+    
+    @ep_times.setter
+    def ep_times(self, arr):
+        self.all[0]['time'] = arr
+
+    @property
+    def ep_durations(self):
+        return self.all[0]['duration']
+    
+    @ep_durations.setter
+    def ep_durations(self, arr):
+        self.all[0]['duration'] = arr
+
+    @property
+    def ep_labels(self):
+        return self.all[0]['label']
+    
+    @ep_labels.setter
+    def ep_labels(self, arr):
+        self.all[0]['label'] = arr
+
+    @property
+    def ep_ids(self):
+        return self.all[0]['id']
+    
+    @ep_ids.setter
+    def ep_ids(self, arr):
+        self.all[0]['id'] = arr
+    
+    @property
+    def ep_stops(self):
+        return self.ep_times + self.ep_durations
+
+    @property
+    def id_to_ind(self):
+        return dict((id,ind) for ind,id in enumerate(self.ep_ids))
+    
```

Comment:
OK for all this setter with properties but if your final goal is multi channel EpochEncoder this will be useless no.
because each channel will in **self.all[0]['label'], self.all[1]['label'], self.all[N]['label']** and so setter will be functions at the end.
Am I missing somethign ?

Reply:
I agree. The multi-channel idea I described [here](https://github.com/NeuralEnsemble/ephyviewer/issues/16#issuecomment-419519065) did not come to me until these properties were already in place. If we implement true multi-channel, these properties will need to change or be removed. I haven't started working on multi-channel since I wanted to stop making changes before you had a chance to review the existing changes.